src/main/java/hudson/matrix/MatrixProject.java

```diff
@@ -660,6 +672,34 @@ public Combination apply(@Nullable List<String> strings) {
         return active;
     }
 
+    /**
+     * Configuration for matrix build
+     */
+    public static class RunConfiguration {
+        public Set<MatrixConfiguration> config;
+        public AxisList axisList;
+    }
+
+    /**
+     * Rebuild project settings and return actual configuration and axis list
+     */
+    /*package*/ RunConfiguration getRunConfiguration(MatrixBuildExecution context) throws IOException {
+        RunConfiguration runConfig = new RunConfiguration();
+        try {
+            buildLock.writeLock().lock();
```

Comment:
As I understand this provides week concurrency guarantee as it protects 2 concurrent dynamic builds but not dynamic build and, say, config update. Is that ok for you?